subprojects/process-services/src/main/java/org/gradle/process/internal/health/memory/CGroupMemoryInfo.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal.health.memory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.Files;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+
+public class CGroupMemoryInfo implements OsMemoryInfo {
+    private static final String CGROUP_MEM_USAGE_FILE = "/sys/fs/cgroup/memory/memory.usage_in_bytes";
+    private static final String CGROUP_MEM_TOTAL_FILE = "/sys/fs/cgroup/memory/memory.limit_in_bytes";
```

Comment:
I've been working on my own version of this, and there is a few of problems with this approach:

1) This only works for cgroups v1. In cgroups v2, there is a "unified hierarchy" where it lives at `/sys/fs/cgroup/memory.max`/`memory.current`.
2) When there is no limit, the number in `memory.limit_in_bytes` (cgroups v1) is a huge number `9223372036854771712`. For `memory.max` (cgroups v2) it is the string `max`. In this case, you can't use these values - you need to read from `/proc/meminfo` to get the total actual system memory.
3) Not sure how much this appears in practice, but it's possible to mount the cgroup filesystems in a different directory to `/sys/fs/cgroup`. The best way to handle this would be:
   i. Read `/etc/mtab` and look for `cgroup2` or `cgroup .... memory` entries
   ii. If these exist, use them
   iii. If they don't exist, try the default `/sys/fs/cgroup` path

Reply:
I can make a PR with the code I have (although might have to wait until Monday, as it's getting late here in London).

Reply:
Thanks for the insight here.  1 and 3 are good points, but also items I think that could be addressed as follow up work (i.e. we could deliver cgroup1 support at the default location in this PR and then add support for cgroup2 later).  I _think_ item 2 is not an issue because when the cgroup1 value is a huge number, we always select the smaller of the cgroup value or the meminfo value.  Assuming we also support cgroups v2 in the future, when it's "max" we would just drop out and not consider the cgroup value (i.e. non-numeric value), defaulting to meminfo.

Reply:
More than happy for this to be merged in as is, it should fix our issues in CI at least. My code unfortunately targets Java 8, rather than 6, so can't immediately get it working when pasted into this repo, but [here it is for posterity](https://github.com/gradle/gradle/compare/gh/pr/17628...CRogers:gradle:callumr/cgroupsv2-memory-info).

Reply:
@ghale is there by any chance any ticket for supporting cgroup2 also?